util/metric/metric.go

```diff
@@ -205,15 +232,29 @@ func (c *Counter) MarshalJSON() ([]byte, error) {
 	return json.Marshal(c.Counter.Count())
 }
 
+// Desc implements prometheus.Metric.
+func (c *Counter) Desc() *prometheus.Desc {
+	return c.desc
+}
+
+// Write implements prometheus.Metric.
+func (c *Counter) Write(met *prometheus_model.Metric) error {
+	met.Gauge = &prometheus_model.Gauge{Value: proto.Float64(float64(c.Count()))}
```

Comment:
Just throwing out there that the only "significant" metric type (Histogram)
has a `.Snapshot` method which is the only thing that would need a lock
around it. If we're not going to generate those snapshots at high frequency
(in which case we must change the API to re-use a snapshot) that should
work out fine. We'll see.

On Fri, Apr 15, 2016 at 11:33 AM marc <notifications@github.com> wrote:

> In util/metric/metric.go
> <https://github.com/cockroachdb/cockroach/pull/6080#discussion_r59892950>:
>
> > @@ -205,15 +232,29 @@ func (c *Counter) MarshalJSON() ([]byte, error) {
> >  	return json.Marshal(c.Counter.Count())
> >  }
> >
> > +// Desc implements prometheus.Metric.
> > +func (c *Counter) Desc() *prometheus.Desc {
> > +	return c.desc
> > +}
> > +
> > +// Write implements prometheus.Metric.
> > +func (c *Counter) Write(met *prometheus_model.Metric) error {
> > +	met.Gauge = &prometheus_model.Gauge{Value: proto.Float64(float64(c.Count()))}
>
> Ah, that might work better, although we are concerned about efficiency as
> well since we expect quite a few metrics, some will lock (as opposed to
> simple atomic values), and may be accessed from all over the code.
>
> I'm using Gauge for now since our underlying Counter implementation allows
> both Inc and Dec. I was going audit the code a bit to make sure we never
> decrements Counter, in which case I could prohibit Dec and switch to the
> prometheus Counter
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6080/files/51ff5f363f8d3fd5f072221915368a4ec40b005e#r59892950>
>
-- 

-- Tobias